### ISSUE TITLE
Unable to freeze graph for LinearClassifier

### ISSUE THREAD

**Issue by dkhmelenko** [tensorflow/tensorflow#18523]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **TensorFlow installed from (source or binary)**: Anaconda
- **TensorFlow version (use command below)**: 1.4.0
- **Python version**: 3.5.4
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**:  N/A
- **GPU model and memory**:  N/A
- **Exact command to reproduce**: freeze_graph for LinearClassifier model

I'm trying to freeze my model using the source https://www.tensorflow.org/mobile/prepare_models. So I could use it in Android app (using Tensorflow Mobile).
I have a model created with `LinearClassifier`. Below is the example of creating it:

```python
gender = tf.feature_column.categorical_column_with_vocabulary_list("gender", ["f", "m"])
goal = tf.feature_column.categorical_column_with_vocabulary_list("goal", ["fit", "lose", "muscle"])
level = tf.feature_column.categorical_column_with_vocabulary_list("level", ["begin", "middle", "advance"])
feature_columns = [gender,goal,level]

input_fn = tf.estimator.inputs.pandas_input_fn(x=x_train, y=y_train, batch_size=100, num_epochs=None, shuffle=True)
model = tf.estimator.LinearClassifier(feature_columns=feature_columns, n_classes=18, model_dir="export")

model.train(input_fn=input_fn, steps=100)
```
The model is successfully created and saved. It works correctly when it's evaluated.
However, I'm getting the following error, when I try to freeze the graph:
```
TypeError: names_to_saveables must be a dict mapping string names to Tensors/Variables. Not a variable: Tensor("linear/linear_model/bias_weights:0", shape=(18,), dtype=float32)
```

I looked on [StackOverflow](https://stackoverflow.com/search?q=names_to_saveables+must+be+a+dict+) and seems this issue happens to many people and there is no answer to that.

I would appreciate any help, because many people are having similar issue.

**Comment by ispirmustafa** [comment]
Hi @dkhmelenko 
Could you please tell us more about how do you freeze the graph?

**Comment by dkhmelenko** [comment]
I'm using the following command for freezing the graph
```
 C:\Users\Dima\Anaconda3\envs\tensorflow\Lib\site-packages\tensorflow\python\tools>python freeze_graph.py --input_graph=export/graph.pbtxt --input_binary=false --input_checkpoint=export/model.ckpt-100 --output_graph=model/frozen_graph.pb --output_node_names=output_node
```
And the output is the following
```
2018-04-16 22:52:21.381508: I C:\tf_jenkins\home\workspace\rel-win\M\windows\PY\35\tensorflow\core\platform\cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX
Traceback (most recent call last):
  File "freeze_graph.py", line 350, in <module>
    app.run(main=main, argv=[sys.argv[0]] + unparsed)
  File "C:\Users\Dima\Anaconda3\envs\tensorflow\lib\site-packages\tensorflow\python\platform\app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "freeze_graph.py", line 249, in main
    FLAGS.saved_model_tags)
  File "freeze_graph.py", line 239, in freeze_graph
    input_meta_graph_def, input_saved_model_dir, saved_model_tags.split(","))
  File "freeze_graph.py", line 127, in freeze_graph_with_def_protos
    saver = saver_lib.Saver(var_list=var_list)
  File "C:\Users\Dima\Anaconda3\envs\tensorflow\lib\site-packages\tensorflow\python\training\saver.py", line 1218, in __init__
    self.build()
  File "C:\Users\Dima\Anaconda3\envs\tensorflow\lib\site-packages\tensorflow\python\training\saver.py", line 1227, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "C:\Users\Dima\Anaconda3\envs\tensorflow\lib\site-packages\tensorflow\python\training\saver.py", line 1263, in _build
    build_save=build_save, build_restore=build_restore)
  File "C:\Users\Dima\Anaconda3\envs\tensorflow\lib\site-packages\tensorflow\python\training\saver.py", line 729, in _build_internal
    saveables = self._ValidateAndSliceInputs(names_to_saveables)
  File "C:\Users\Dima\Anaconda3\envs\tensorflow\lib\site-packages\tensorflow\python\training\saver.py", line 632, in _ValidateAndSliceInputs
    variable)
TypeError: names_to_saveables must be a dict mapping string names to Tensors/Variables. Not a variable: Tensor("linear/linear_model/bias_weights:0", shape=(18,), dtype=float32)

(tensorflow) C:\Users\Dima\Anaconda3\envs\tensorflow\Lib\site-packages\tensorflow\python\tools>
```

The files `graph.pbtxt` and `model.ckpt-100` are in attachment.
[checkpoint.zip](https://github.com/tensorflow/tensorflow/files/1917237/checkpoint.zip)

**Comment by ispirmustafa** [comment]
Pete, could you please take a look?

**Comment by dkhmelenko** [comment]
@ispirmustafa @petewarden are there any updates on this issue?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Madhusakth** [comment]
I am facing a similar issue. I gave the following as command in Anaconda prompt: 
freeze_graph --input_graph=C:\Users\msakthi\Desktop\Speech_rec\speech_commands\9wordsmodel\conv.pb --input_checkpoint=C:\Users\msakthi\Desktop\Speech_rec\speech_commands\9wordsmodel\conv.ckpt-18000 --output_graph=C:\Users\msakthi\Desktop\Speech_rec\speech_commands\9wordsmodel\conv_graph.pb --output_node_names=MatMul --input_binary=true

yet, I get an error saying:
TypeError: names_to_saveables must be a dict mapping string names to Tensors/Variables. Not a variable: Tensor("Variable:0", shape=(20, 8, 1, 64), dtype=float32).

If you could help me with this issue, great! 

Thanks in advance.

**Comment by dkhmelenko** [comment]
The issue still exists for Tensorflow 1.8.0

**Comment by petewarden** [comment]
This error indicates a general issue with trying to freeze a graph that doesn't match the checkpoint. I can't tell from the code how you've saved out the graph, but the most reliable way to freeze is to use a SavedModel folder where you can be sure the graph and checkpoints do match, and you only have to pass in a single folder location.

**Comment by dkhmelenko** [comment]
@petewarden Thank you for your response.

I've tried a couple of the approaches and none of them worked for me:
1. I tried to freeze the graph using the function `model.export_savedmodel`. Then I've got a folder with the files `saved_model.pb`, `variables.data-00000-of-00001` and `variables.index`. But there is no checkpoint file in this case. What should be done in this case?

2. I tried also to freeze the graph using the argument `model_dir` on creating `LinearClassifier`. In that case, I've got the files `graph.pbtxt`, `checkpoint` and a couple of `model.ckpt` files. But still no success.

My code is pretty straightforward and can be found here https://github.com/dkhmelenko/fitness-ml/blob/master/simple/Fitness%20Classification.ipynb

I would appreciate any kind of help from your side. Thank you in advance.

**Comment by harsh-agar** [comment]
@dkhmelenko  did you find any solution to the mentioned error?

**Comment by zhangjf23** [comment]
@dkhmelenko did you find any solution to the mentioned error?

**Comment by OzzieFZI** [comment]
I have the same issue in v1.9. In my case, I am trying to get a .tflite model to serve from Firebase.
The other method I tried (which has been mentioned above) is creating a SavedModel then attempting to convert using 
`tf.contrib.lite.TocoConverter.from_saved_model(path_to_savedModel)`
but then I get a different error:
**ValueError: Tensors input_example_tensor:0 not known type tf.string**

**Comment by dkhmelenko** [comment]
@harsh-agar @zhangjf23 sorry, I didnt find any solution so far, I'm still facing the issue

**Comment by gargn** [comment]
@OzzieFZI The issue you listed has been fixed in head. To use the nightly build either use "pip install --upgrade tf_nightly" or follow the instructions [here](https://www.tensorflow.org/install/install_sources).

**Comment by gargn** [comment]
@OzzieFZI The list of supported ops are available [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/g3doc/tf_ops_compatibility.md). The two recommended solutions are:
1. Add a custom op as described [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/g3doc/custom_operators.md).
2. Only use the ops that are currently supported by TFLite in your model.
For any additional issues with your model, please make another bug so we can track issues separately.

@dkhmelenko Could you attach a SavedModel for your model?

**Comment by dkhmelenko** [comment]
@gargn 
Here is SavedModel as a zip. It contains `saved_model.pbtxt` and `variables` directory.
[SavedModel.zip](https://github.com/tensorflow/tensorflow/files/2229710/SavedModel.zip)

**Comment by gargn** [comment]
@dkhmelenko I looked into the issue further and there seem to be some alternative ways that you can run the code using your SavedModel.

1. If you intend to run your code through TensorFlow Lite then you can use either [`TocoConverter`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/toco/g3doc/python_api.md#basic-savedmodel) (Python API) or [`tflite_convert`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/toco/g3doc/cmdline_examples.md#savedmodel) (command line tool) with your SavedModel. This should be working starting from TensorFlow 1.9. If not, it works on the nightly build. However, this results in the error mentioned above indicating there are some operators that aren't supported. The solution to that error is mentioned in the comment below.

2. If you just want to run `freeze_graph.py` then you should be able to run it with a SavedModel. The current error you are getting is coming from the code that loads the checkpoints. The command line below should work with your code. The values for `output_node_names` came from the SignatureDef within your SavedModel.

```
python freeze_graph.py \
--input_saved_model_dir=SAVED_MODEL_DIR \
--output_node_names=linear/head/Tile,linear/head/predictions/probabilities \
--output_graph=/tmp/frozen_graph.pb
```

Let me know if one of these solutions suffices. In the meantime I'm looking for a solution to fix the problem in `freeze_graph.py`.

**Comment by gargn** [comment]
After some investigation, it appears the only way to freeze models with partitioned variables is using SavedModels. The error message has been [updated](https://github.com/tensorflow/tensorflow/commit/37594fd0945061aee9f4d5f6ba2aa8c4b360697c) to state this clearly.

I am closing the issue since there is working solution and an improved error message. Please reopen if there are any additional issues relating to freezing your model.

**Comment by chrisrapson** [comment]
I'm having the same problem. When I tried to use the SavedModels option, I got the following error: `IOError: SavedModel file does not exist at: /path/I/supplied/as/argument`

Maybe I am misunderstanding what this argument should be. I thought it would be the path to the folder which has the ascii graph and trios of ckpt files (index, meta, data) in it? The same one that I provide as the `--logdir` argument to tensorboard?
